pkg/services/dashboards/database/altsearch.go

```diff
@@ -0,0 +1,188 @@
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, 0, parent_uid, title FROM folder WHERE MATCH(title) AGAINST({{ .Arg .Text }})
+{{ end }}
+{{- define "text-search-legacy" -}}
+SELECT org_id, CASE WHEN is_folder THEN 'folder' ELSE 'dashboard' END as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND NOT is_folder
+{{ end }}
+
+SELECT org_id, kind, uid, entity.id, title, dashboard_tag.term as term, folder_uid, folder_title FROM (
+	SELECT DISTINCT
+		entity.org_id, entity.kind, entity.uid, entity.id, entity.title, f1.uid as folder_uid, f1.title as folder_title FROM (
+		{{- template "text-search-legacy" . -}}
+	) AS entity
+	LEFT JOIN folder f1 ON (entity.parent_uid = f1.uid AND entity.org_id = f1.org_id)
+	{{ if ne .UserID 0 }}   {{- /* If a user is not admin: need to filter by permissions */ -}}
+	LEFT JOIN folder f2 ON (f1.parent_uid     = f2.uid AND f1.org_id     = f2.org_id)
+	LEFT JOIN folder f3 ON (f2.parent_uid     = f3.uid AND f2.org_id     = f3.org_id)
+	LEFT JOIN folder f4 ON (f3.parent_uid     = f4.uid AND f3.org_id     = f4.org_id)
+	INNER JOIN permission p ON 
+	p.role_id IN (
+		SELECT ur.role_id FROM user_role AS ur    WHERE ur.user_id = {{ .Arg .UserID }} AND (ur.org_id = {{ .Arg .OrgID }} OR ur.org_id = 0)
+		{{ if .TeamIDs }}
+		UNION
+		SELECT tr.role_id FROM team_role AS tr    WHERE tr.team_id IN (
+			{{- range $i, $el := .TeamIDs -}}
+			{{- if $i -}},{{- end -}}
+			{{- $.Arg $el -}}
+			{{- end -}}
+		) AND tr.org_id = {{ .Arg .OrgID }}
+		{{ end }}
+		UNION
+		SELECT br.role_id FROM builtin_role AS br WHERE br.role IN ({{ .Arg .Role }}) AND (br.org_id = {{ .Arg .OrgID }} OR br.org_id = 0)
+	) AND (
+		{{- /* This permission check relies on the optimised (indexed) split scope columns in the permission table */ -}}
+		(
+			p.kind = 'dashboards' AND
+			p.action = 'dashboards:read' AND
+			p.identifier IN (entity.uid, f4.uid, f3.uid, f2.uid, f1.uid)
+		) OR (
+			p.kind = 'folders' AND
+			p.action = 'folders:read' AND
+			p.identifier IN (f4.uid, f3.uid, f2.uid, f1.uid, entity.uid)
+		)
+	)
+	{{ end }}
+	ORDER BY entity.title ASC
+	LIMIT {{ .Arg .Limit }}
+) AS entity
+LEFT JOIN dashboard_tag ON dashboard_tag.dashboard_id = entity.id
+`))
+
+func (d *dashboardStore) findDashboards(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	// Only handle non-empty search queries
+	if query.Title == "" {
+		return nil, nil
+	}
+
+	// Get OrgID, Team IDs and UserID (zero for admins)
+	orgID, userID, teamIDs := query.OrgId, int64(0), []int64{}
+	if orgID == 0 {
+		orgID = query.SignedInUser.GetOrgID()
+	}
+	if !query.SignedInUser.GetIsGrafanaAdmin() {
+		namespaceID, identifier := query.SignedInUser.GetNamespacedID()
+		switch namespaceID {
+		case identity.NamespaceUser, identity.NamespaceServiceAccount:
+			userID, _ = identity.IntIdentifier(namespaceID, identifier)
+		}
```

Comment:
[question] This means that anonymous would be treated as admin? From what I see, having id 0 would be admin in the query since all permissions checks are behind `{{ if ne .UserID 0 }}`, right? Should we fail in a `default` case just to make sure?